From 770799128d4cf0b2908f8cd15d935fc6cc24397a Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?V=C3=ADt=20Ondruch?= <vondruch@redhat.com>
Date: Fri, 8 Feb 2013 18:20:50 +0100
Subject: [PATCH] Add configuration arch specific dir flag.

---
 configure.in | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/configure.in b/configure.in
index 7604bb8..418b0cb 100644
--- a/configure.in
+++ b/configure.in
@@ -3186,6 +3191,10 @@ AC_ARG_WITH(rubylibprefix,
 rubylibdir='${rubylibprefix}/${ruby_version}'
 rubyarchdir=${multiarch+'${rubyarchprefix}/${ruby_version}'}${multiarch-'${rubylibdir}/${arch}'}
 
+AC_ARG_WITH(archdir,
+     AS_HELP_STRING([--with-archdir=DIR], [architecture specific ruby libraries [[LIBDIR/RUBY_BASE_NAME/ARCH]]]),
+            [rubyarchdir=$withval])
+
 rubyarchprefix=${multiarch+'${archlibdir}/${RUBY_BASE_NAME}'}${multiarch-'${rubylibprefix}/${arch}'}
 AC_ARG_WITH(rubyarchprefix,
 	    AS_HELP_STRING([--with-rubyarchprefix=DIR],
-- 
1.8.1

