diff -pruN ruby-1.8.6-p111.orig/ext/socket/socket.c ruby-1.8.6-p111/ext/socket/socket.c
--- ruby-1.8.6-p111.orig/ext/socket/socket.c	2007-05-23 00:08:43.000000000 +0900
+++ ruby-1.8.6-p111/ext/socket/socket.c	2008-02-19 11:24:22.000000000 +0900
@@ -893,10 +893,10 @@ port_str(port, pbuf, len)
 }
 
 #ifndef NI_MAXHOST
-# define 1025
+# define NI_MAXHOST 1025
 #endif
 #ifndef NI_MAXSERV
-# define 32
+# define NI_MAXSERV 32
 #endif
 
 static struct addrinfo*
