diff -ruN ruby-1.8.3.orig/mkconfig.rb ruby-1.8.3/mkconfig.rb
--- ruby-1.8.3.orig/mkconfig.rb	2005-09-21 15:10:57.000000000 +0900
+++ ruby-1.8.3/mkconfig.rb	2005-09-21 15:17:32.000000000 +0900
@@ -74,7 +74,7 @@
 
 drive = File::PATH_SEPARATOR == ';'
 
-prefix = '/lib/ruby/' + RUBY_VERSION.sub(/\.\d+$/, '') + '/' + RUBY_PLATFORM
+prefix = '/lib64/ruby/' + RUBY_VERSION.sub(/\.\d+$/, '') + '/' + RUBY_PLATFORM
 print "  TOPDIR = File.dirname(__FILE__).chomp!(#{prefix.dump})\n"
 print "  DESTDIR = ", (drive ? "TOPDIR && TOPDIR[/\\A[a-z]:/i] || " : ""), "'' unless defined? DESTDIR\n"
 print "  CONFIG = {}\n"
diff -ruN ruby-1.8.3.orig/configure.in ruby-1.8.3/configure.in
--- ruby-1.8.3.orig/configure.in	2005-09-07 08:22:56.000000000 +0900
+++ ruby-1.8.3/configure.in	2005-09-26 17:46:53.000000000 +0900
@@ -1424,15 +1424,15 @@
     RUBY_LIB_PREFIX="/lib/ruby"
     ;;
   *)
-    RUBY_LIB_PREFIX="${prefix}/lib/ruby"
+    RUBY_LIB_PREFIX="${libdir}/ruby"
     ;;
 esac
 RUBY_LIB_PATH="${RUBY_LIB_PREFIX}/${MAJOR}.${MINOR}"
 
 AC_ARG_WITH(sitedir,
-	    [  --with-sitedir=DIR      site libraries in DIR [PREFIX/lib/ruby/site_ruby]],
+	    [  --with-sitedir=DIR      site libraries in DIR [LIBDIR/ruby/site_ruby]],
             [sitedir=$withval],
-            [sitedir='${prefix}/lib/ruby/site_ruby'])
+            [sitedir='${libdir}/ruby/site_ruby'])
 SITE_DIR="`eval \"echo ${sitedir}\"`"
 case "$target_os" in
   cygwin*|mingw*|*djgpp*|os2-emx*)
