diff -ruN ruby-1.8.4.orig/configure.in ruby-1.8.4/configure.in
--- ruby-1.8.4.orig/configure.in	2005-11-24 21:07:18.000000000 +0900
+++ ruby-1.8.4/configure.in	2006-02-06 12:51:58.000000000 +0900
@@ -437,7 +437,7 @@
 AC_REPLACE_FUNCS(dup2 memmove strcasecmp strncasecmp strerror strftime\
 		 strchr strstr strtoul crypt flock vsnprintf\
 		 isnan finite isinf hypot acosh erf)
-AC_CHECK_FUNCS(fmod killpg wait4 waitpid syscall chroot fsync getcwd\
+AC_CHECK_FUNCS(fmod killpg wait4 waitpid syscall chroot fsync getcwd eaccess\
 	      truncate chsize times utimes fcntl lockf lstat symlink link\
 	      readlink setitimer setruid seteuid setreuid setresuid\
 	      setproctitle setrgid setegid setregid setresgid issetugid pause\
diff -ruN ruby-1.8.4.orig/file.c ruby-1.8.4/file.c
--- ruby-1.8.4.orig/file.c	2005-12-21 18:20:15.000000000 +0900
+++ ruby-1.8.4/file.c	2006-02-06 12:52:48.000000000 +0900
@@ -849,6 +849,7 @@
 #  define S_IXUGO		(S_IXUSR | S_IXGRP | S_IXOTH)
 #endif
 
+#ifndef HAVE_EACCESS
 int
 eaccess(path, mode)
      const char *path;
@@ -887,6 +888,7 @@
     return access(path, mode);
 #endif
 }
+#endif
 
 
 /*
diff -ruN ruby-1.8.4.orig/intern.h ruby-1.8.4/intern.h
--- ruby-1.8.4.orig/intern.h	2005-11-23 11:06:00.000000000 +0900
+++ ruby-1.8.4/intern.h	2006-02-06 12:52:54.000000000 +0900
@@ -221,7 +221,6 @@
 void rb_thread_atfork _((void));
 VALUE rb_funcall_rescue __((VALUE, ID, int, ...));
 /* file.c */
-int eaccess _((const char*, int));
 VALUE rb_file_s_expand_path _((int, VALUE *));
 VALUE rb_file_expand_path _((VALUE, VALUE));
 void rb_file_const _((const char*, VALUE));
diff -ruN ruby-1.8.4.orig/missing.h ruby-1.8.4/missing.h
--- ruby-1.8.4.orig/missing.h	2005-06-13 01:58:41.000000000 +0900
+++ ruby-1.8.4/missing.h	2006-02-06 12:53:29.000000000 +0900
@@ -39,6 +39,10 @@
 extern int dup2 _((int, int));
 #endif
 
+#ifndef HAVE_EACCESS
+extern int eaccess(const char*, int);
+#endif
+
 #ifndef HAVE_FINITE
 extern int finite _((double));
 #endif
