Index: ruby_1_8/ruby.h
===================================================================
--- ruby_1_8/ruby.h	(revision 13310)
+++ ruby_1_8/ruby.h	(revision 13311)
@@ -374,6 +374,11 @@
     int iter_lev;
     VALUE ifnone;
 };
+#define RHASH_TBL(h) (RHASH(h)->tbl)
+#define RHASH_ITER_LEV(h) (RHASH(h)->iter_lev)
+#define RHASH_IFNONE(h) (RHASH(h)->ifnone)
+#define RHASH_SIZE(h) (RHASH(h)->tbl->num_entries)
+#define RHASH_EMPTY_P(h) (RHASH_SIZE(h) == 0)
 
 struct RFile {
     struct RBasic basic;
