--- ruby-1.8.6-p369/lib/rdoc/ri/ri_options.rb.gemmulti	2007-02-13 08:01:19.000000000 +0900
+++ ruby-1.8.6-p369/lib/rdoc/ri/ri_options.rb	2009-10-14 23:51:09.000000000 +0900
@@ -63,7 +63,7 @@
 
         [ "--gems",         nil,    nil,
           "Include documentation from Rubygems:\n  " +
-          (RI::Paths::GEMDIRS ? "#{Gem.path}/doc/*/ri" :
+          (RI::Paths::GEMDIRS ? Gem.path.map {|path| "#{path}/doc/*/ri" }.join(', ') :
                                 "No Rubygems ri found.") ],
                                                            
         [ "--format",       "-f",   "<name>",
@@ -136,7 +136,11 @@
           RI::Paths::HOMEDIR
         ]
 
-        directories << "#{Gem.path}/doc/*/ri" if RI::Paths::GEMDIRS
+        if RI::Paths::GEMDIRS
+          Gem.path.each {|gempath|
+            directories << "#{gempath}/doc/*/ri"
+          }
+        end
 
         directories = directories.join("\n    ")
 
--- ruby-1.8.6-p369/lib/rdoc/ri/ri_paths.rb.gemmulti	2007-02-13 08:01:19.000000000 +0900
+++ ruby-1.8.6-p369/lib/rdoc/ri/ri_paths.rb	2009-10-15 00:02:05.000000000 +0900
@@ -44,7 +44,7 @@
 
     begin
       require 'rubygems'
-      GEMDIRS = Dir["#{Gem.path}/doc/*/ri"]
+      GEMDIRS = Gem.path.map {|path| Dir["#{path}/doc/*/ri"]}.flatten
       GEMDIRS.each { |path| RI::Paths::PATH << path }
     rescue LoadError
       GEMDIRS = nil
