From 2f03b54b63043aadce9e95b83e98b29712bad21f Mon Sep 17 00:00:00 2001
From: Eric Hodel <drbrain@segment7.net>
Date: Mon, 6 Jan 2014 14:30:07 -0800
Subject: [PATCH] Use source_paths for fallback rdoc plugin

Fixes #758
---
 lib/rubygems/rdoc.rb | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/rubygems/rdoc.rb b/lib/rubygems/rdoc.rb
index 52249dc..633bd89 100644
--- a/lib/rubygems/rdoc.rb
+++ b/lib/rubygems/rdoc.rb
@@ -193,7 +193,7 @@ def generate
     ::RDoc::Parser::C.reset
 
     args = @spec.rdoc_options
-    args.concat @spec.require_paths
+    args.concat @spec.source_paths
     args.concat @spec.extra_rdoc_files
 
     case config_args = Gem.configuration[:rdoc]
-- 
1.8.5.1

