diff -pruN ruby-1.8.6.orig/ext/bigdecimal/bigdecimal.c ruby-1.8.6/ext/bigdecimal/bigdecimal.c
--- ruby-1.8.6.orig/ext/bigdecimal/bigdecimal.c	2007-02-27 20:51:55.000000000 +0900
+++ ruby-1.8.6/ext/bigdecimal/bigdecimal.c	2007-03-15 21:14:32.000000000 +0900
@@ -2167,12 +2167,14 @@ VpGetDoubleNegZero(void) /* Returns the 
     return nzero;
 }
 
+#if 0
 VP_EXPORT int
 VpIsNegDoubleZero(double v)
 {
     double z = VpGetDoubleNegZero();
     return MemCmp(&v,&z,sizeof(v))==0;
 }
+#endif
 
 VP_EXPORT int
 VpException(unsigned short f, const char *str,int always)
diff -pruN ruby-1.8.6.orig/parse.y ruby-1.8.6/parse.y
--- ruby-1.8.6.orig/parse.y	2007-03-03 16:30:46.000000000 +0900
+++ ruby-1.8.6/parse.y	2007-03-15 21:15:58.000000000 +0900
@@ -4511,6 +4511,7 @@ rb_node_newnode(type, a0, a1, a2)
     return n;
 }
 
+#if 0
 static enum node_type
 nodetype(node)			/* for debug */
     NODE *node;
@@ -4524,6 +4525,7 @@ nodeline(node)
 {
     return nd_line(node);
 }
+#endif
 
 static NODE*
 newline_node(node)
diff -pruN ruby-1.8.6.orig/regex.c ruby-1.8.6/regex.c
--- ruby-1.8.6.orig/regex.c	2007-02-27 20:51:55.000000000 +0900
+++ ruby-1.8.6/regex.c	2007-03-15 21:16:52.000000000 +0900
@@ -998,6 +998,7 @@ print_partial_compiled_pattern(start, en
 }
 
 
+#if 0
 static void
 print_compiled_pattern(bufp)
      struct re_pattern_buffer *bufp;
@@ -1006,6 +1007,7 @@ print_compiled_pattern(bufp)
 
   print_partial_compiled_pattern(buffer, buffer + bufp->used);
 }
+#endif
 
 static char*
 calculate_must_string(start, end)
@@ -2586,6 +2588,7 @@ insert_jump_n(op, from, to, current_end,
 
    If you call this function, you must zero out pending_exact.  */
 
+#if 0
 static void
 insert_op(op, there, current_end)
      int op;
@@ -2599,6 +2602,7 @@ insert_op(op, there, current_end)
 
   there[0] = (char)op;
 }
+#endif
 
 
 /* Open up space at location THERE, and insert operation OP followed by
